oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,381 @@
+    return headers;
+  }
+
+  private AwsDates getDates() {
+    if (additionalHeaders.containsKey("date")) {
+      return AwsDates.fromDateHeader(additionalHeaders.get("date"));
+    }
+    if (additionalHeaders.containsKey("x-amz-date")) {
+      return AwsDates.fromXAmzDate(additionalHeaders.get("x-amz-date"));
+    }
+    return AwsDates.generateXAmzDate();
+  }
+
+  private static byte[] sign(byte[] key, byte[] value) {
+    try {
+      String algorithm = "HmacSHA256";
+      Mac mac = Mac.getInstance(algorithm);
+      mac.init(new SecretKeySpec(key, algorithm));
+      return mac.doFinal(value);
+    } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+      throw new RuntimeException("Failed to calculate the AWS V4 Signature.", e);
+    }
+  }
+
+  private static String getHexEncodedSha256Hash(byte[] bytes) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("SHA-256");
+      return BaseEncoding.base16().lowerCase().encode(digest.digest(bytes));
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException("Failed to compute SHA-256 hash.", e);
+    }
+  }
+
+  static Builder newBuilder(
+      AwsSecurityCredentials awsSecurityCredentials, String httpMethod, String url, String region) {
+    return new Builder(awsSecurityCredentials, httpMethod, url, region);
+  }
+
+  static class Builder {
+
+    private AwsSecurityCredentials awsSecurityCredentials;
+    private String httpMethod;
+    private String url;
+    private String region;
+
+    @Nullable private String requestPayload;
+    @Nullable private Map<String, String> additionalHeaders;
+
+    private Builder(
+        AwsSecurityCredentials awsSecurityCredentials,
+        String httpMethod,
+        String url,
+        String region) {
+      this.awsSecurityCredentials = awsSecurityCredentials;
+      this.httpMethod = httpMethod;
+      this.url = url;
+      this.region = region;
+    }
+
+    Builder setRequestPayload(String requestPayload) {
+      this.requestPayload = requestPayload;
+      return this;
+    }
+
+    Builder setAdditionalHeaders(Map<String, String> additionalHeaders) {
+      this.additionalHeaders = additionalHeaders;
+      return this;
+    }
+
+    AwsRequestSigner build() {
+      return new AwsRequestSigner(
+          awsSecurityCredentials, httpMethod, url, region, requestPayload, additionalHeaders);
+    }
+  }
+
+  static final class AwsDates {
+
+    private static final String X_AMZ_DATE_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
+    private static final String CUSTOM_DATE_FORMAT = "E, dd MMM yyyy HH:mm:ss z";
+
+    private String originalDate;
+    private String amzDate;
+    private String formattedDate;
+
+    private AwsDates(String originalDate, String amzDate, String formattedDate) {
+      this.originalDate = checkNotNull(originalDate);
+      this.amzDate = checkNotNull(amzDate);
+      this.formattedDate = checkNotNull(formattedDate);
+    }
+
+    static AwsDates fromXAmzDate(String xAmzDate) {
```

Comment:
The date isn't validated. Do we know its good here?

Reply:
Validated in the Builder now.